From: SeongJae Park <sj@kernel.org>
Date: Sat, 7 Dec 2024 12:15:40 -0800
Subject: [PATCH] ==== sz_ops_filter_passed stats ====

mm/damon: enable page level properties based access pattern monitoring

TL; DR
======

This patch series eanble access monitoring based on page level
properties including their anonymousness, belonging cgroups and
pg-young-ness, by extending DAMOS stats and regions walk features with
region-internal DAMOS filters.

Background
==========

DAMON fundametally provides only access pattern information in region
granularity.  For some use cases, fixed and fine granularity information
based on non access pattern properties can be useful, though.  For
example, on systems having fast swap devices and slow storage devices,
DAMOS-based proactive reclaim need to be applied differently for
anonymous pages and file-backed pages.

DAMOS makes it possible via DAMOS filters that work in page level
properties including anonymousness, belonging cgroups, and
pg-young-ness.  Having only the fine-grained inforamtion without making
system operation behavior changes can be useful for tuning or monitoring
purposes.  But currently DAMON provides no way for that.

Idea
====

DAMOS has initially developed for only access-aware system operations.
But, efficient acces monitoring results querying is yet another major
usage of today's DAMOS.  DAMOS stats and regions walk, which exposes
accumulated counts and per-region monitoring results that filtered by
DAMOS parameters including target access pattern and quotas, are the key
features for that usage.  Special DAMOS action, DAMOS_STAT, was also
introduced to support doing only monitoring without making any system
operation behavioral changes.

Hence, the fine-grained information that already available for
access-aware system operation can be passed to users by connecting the
region-internal DAMOS filters with stats and regions walk features.

Design
======

Update the interface of DAMON operations set layer, which contains
region-internal DAMOS filters implementation, to report back the amount
of memory that passed the region-internal DAMOS filters to the core
layer.  On the core layer, account the operations set layer reported
stat with a per-scheme accumulated stat.  Also, pass the stat to regions
walk.  In this way, DAMON API users can efficiently get the fine-grained
information.

For the user-space, make DAMON sysfs interface collects the information
using DAMON core API, and expose those to new per-scheme stats file and
per-DAMOS-tried region properties file.

Practical Usages
================

With this patch series, DAMON users can query how many bytes of regions
of specific access temperature is backed by pages of specific type.  The
type can be any of DAMOS filter-supporting one, including anonymousness,
belonging cgroups, and pg-young-ness.  For example, users can visualize
access hotness-based page granulairty histogram for different cgroups,
backing content type, or youngness.  In future, it could be extended to
more types such as whether it is THP, position on LRU lists, etc.  This
can be useful for estimating benefits of a new or an existing
access-aware system optimizations without really committing the changes.

Patches Sequence
================

The patches are separated in four sequences.

First four patches add adding documentation of background knowledge.

Following three patches change the operations set layer interface to
report back the region-internal filter passed memory size, and make the
operations set implementations support the changed symantic.

Following five patches implement per-scheme accumulated stat for
region-internal filter-passed memory size on core API, DAMON sysfs
interface.  First two patches of those are for code change, and
following three patches are for documentation.

Finally, six patches implementing per-region region-internal
filter-passed memory size follows.  First three patches of these are for
code change of preparation, API, and DAMON sysfs interface parts.  The
final three patches are for documentation.

Revision History
================

Changes from RFC
(https://lore.kernel.org/20241219040327.61902-1-sj@kernel.org)
- Fix kernel-doc undocumented parameter
  (https://lore.kernel.org/oe-kbuild-all/202412191225.f6bEMRT2-lkp@intel.com/)
- Drop walk_fn() invocation sequence and regions walk documentation
  - Those are moved to damos_call() intro patch series

Signed-off-by: SeongJae Park <sj@kernel.org>
---
 damon_meta_changes/D7jYt0Oq | 0
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 damon_meta_changes/D7jYt0Oq

diff --git a/damon_meta_changes/D7jYt0Oq b/damon_meta_changes/D7jYt0Oq
new file mode 100644
index 000000000000..e69de29bb2d1
-- 
2.39.5

